File: app/src/main/java/name/monwf/customiuizer/MainActivity.java
package name.monwf.customiuizer;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

import io.github.libxposed.service.RemotePreferences;
import io.github.libxposed.service.XposedService;
import io.github.libxposed.service.XposedServiceHelper;
import name.monwf.customiuizer.mods.GlobalActions;
import name.monwf.customiuizer.utils.AppHelper;
import name.monwf.customiuizer.utils.Helpers;

public class MainActivity extends AppCompatActivity {

    MainFragment mainFrag = null;
    SharedPreferences.OnSharedPreferenceChangeListener prefsChanged;

    @Override
    protected void attachBaseContext(Context base) {
        try {
            super.attachBaseContext(AppHelper.getLocaleContext(base));
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (AppHelper.remotePrefs == null) {
            XposedServiceHelper.registerListener(new XposedServiceHelper.OnServiceListener() {
                public void onServiceBind(XposedService service) {
                    AppHelper.moduleActive = true;
                    AppHelper.remotePrefs = (RemotePreferences) service.getRemotePreferences(AppHelper.prefsName + "_remote");
                }
                public void onServiceDied(XposedService service) {
                    AppHelper.moduleActive = false;
                    AppHelper.remotePrefs = null;
                }
            });
        }

//		Helpers.updateNewModsMarking(this);

        Toolbar myToolbar = findViewById(R.id.mainActionBar);
        setSupportActionBar(myToolbar);
        if (savedInstanceState != null) {
            mainFrag = (MainFragment) getSupportFragmentManager().getFragment(savedInstanceState, "mainFrag");
        }
        else if (mainFrag == null) {
            mainFrag = new MainFragment();
            getSupportFragmentManager().beginTransaction().setReorderingAllowed(true).replace(R.id.fragment_container, mainFrag).commit();
        }

        HashSet<String> ignoreKeys = new HashSet<>();
        ignoreKeys.add("pref_key_miuizer_locale");
        ignoreKeys.add("pref_key_miuizer_launchericon");
        ignoreKeys.add("pref_key_miuizer_synced_from_lsposed");
        prefsChanged = new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
                if (AppHelper.remotePrefs == null) return;
//                AppHelper.log("app changed key: " + key);
                if (key == null) {
                    RemotePreferences.Editor prefEdit = AppHelper.remotePrefs.edit();
                    for (String remoteKey:AppHelper.remotePrefs.getAll().keySet()) {
                        prefEdit.remove(remoteKey);
                    }
                    prefEdit.apply();
                    return ;
                }
                if (ignoreKeys.contains(key)) return;
                Object val = sharedPreferences.getAll().get(key);
                RemotePreferences.Editor prefEdit = AppHelper.remotePrefs.edit();
                if (val == null) {
                    prefEdit.remove(key);
                }
                else if (val instanceof Boolean) {
                    prefEdit.putBoolean(key, (Boolean) val);
                }
                else if (val instanceof Float) {
                    prefEdit.putFloat(key, (Float)val);
                }
                else if (val instanceof Integer) {
                    prefEdit.putInt(key, (Integer) val);
                }
                else if (val instanceof Long) {
                    prefEdit.putLong(key, (Long) val);
                }
                else if (val instanceof String) {
                    prefEdit.putString(key, (String) val);
                }
                else if (val instanceof Set<?>) {
                    prefEdit.putStringSet(key, (Set<String>) val);
                }
                prefEdit.apply();
            }
        };

        AppHelper.appPrefs.registerOnSharedPreferenceChangeListener(prefsChanged);
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        getSupportFragmentManager().putFragment(savedInstanceState, "mainFrag", mainFrag);
        super.onSaveInstanceState(savedInstanceState);
    }

    @SuppressLint("ApplySharedPref")
    protected void onDestroy() {
        try {
            if (prefsChanged != null) AppHelper.appPrefs.unregisterOnSharedPreferenceChangeListener(prefsChanged);
        } catch (Throwable t) {
            t.printStackTrace();
        }
        super.onDestroy();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fragment_container);
            if (fragment == null) {
                finish();
                return true;
            }
            if (fragment instanceof MainFragment) {
                finish();
            }
            else {
                ((SubFragment)fragment).finish();
            }
            return true;
        }
        else if (item.getItemId() == R.id.resetsettings) {
            showResetSettingsDialog();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void showResetSettingsDialog() {
        AlertDialog.Builder alert = new AlertDialog.Builder(this);
        alert.setTitle(R.string.reset_settings);
        alert.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                AppHelper.appPrefs.edit().clear().commit();
                AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
                builder.setTitle(R.string.reset_settings_done);
                builder.setCancelable(true);
                builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton){
                        int pid = android.os.Process.myPid();
                        android.os.Process.killProcess(pid);
                    }
                });
                builder.show();
            }
        });
        alert.setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {}
        });
        alert.show();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_MENU) {
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (grantResults.length == 0) {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
            return;
        }

        switch (requestCode) {
            case Helpers.REQUEST_PERMISSIONS_WIFI:
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    if (shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION))
                        Toast.makeText(this, R.string.permission_scan, Toast.LENGTH_LONG).show();
                    else
                        Toast.makeText(this, R.string.permission_permanent, Toast.LENGTH_LONG).show();
                }
                break;
            case Helpers.REQUEST_PERMISSIONS_BLUETOOTH:
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    if (shouldShowRequestPermissionRationale(Manifest.permission.BLUETOOTH_CONNECT))
                        Toast.makeText(this, R.string.permission_scan, Toast.LENGTH_LONG).show();
                    else
                        Toast.makeText(this, R.string.permission_permanent, Toast.LENGTH_LONG).show();
                }
                break;
            case Helpers.REQUEST_PERMISSIONS_REPORT:
                Toast.makeText(this, ":(", Toast.LENGTH_SHORT).show();
                break;
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }
}

File: app/src/main/java/tv/withaibuild/customiuizer/MainActivity.kt
package tv.withaibuild.customiuizer

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Process
import android.view.KeyEvent
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import io.github.libxposed.service.RemotePreferences
import io.github.libxposed.service.XposedService
import io.github.libxposed.service.XposedServiceHelper
import tv.withaibuild.customiuizer.utils.AppHelper
import tv.withaibuild.customiuizer.utils.Helpers

class MainActivity : AppCompatActivity() {

    private var mainFrag: MainFragment? = null
    private var prefsChanged: SharedPreferences.OnSharedPreferenceChangeListener? = null

    override fun attachBaseContext(base: Context) {
        try {
            super.attachBaseContext(AppHelper.getLocaleContext(base) ?: base)
        } catch (t: Throwable) {
            t.printStackTrace()
            super.attachBaseContext(base)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (AppHelper.remotePrefs == null) {
            XposedServiceHelper.registerListener(object : XposedServiceHelper.OnServiceListener {
                override fun onServiceBind(service: XposedService) {
                    AppHelper.moduleActive = true
                    AppHelper.remotePrefs = service.getRemotePreferences(AppHelper.prefsName + "_remote") as RemotePreferences
                }

                override fun onServiceDied(service: XposedService) {
                    AppHelper.moduleActive = false
                    AppHelper.remotePrefs = null
                }
            })
        }

        val myToolbar = findViewById<Toolbar>(R.id.mainActionBar)
        setSupportActionBar(myToolbar)
        if (savedInstanceState != null) {
            mainFrag = supportFragmentManager.getFragment(savedInstanceState, "mainFrag") as? MainFragment
        } else if (mainFrag == null) {
            mainFrag = MainFragment()
            supportFragmentManager.beginTransaction()
                .setReorderingAllowed(true)
                .replace(R.id.fragment_container, mainFrag!!)
                .commit()
        }

        val ignoreKeys = HashSet<String>().apply {
            add("pref_key_miuizer_locale")
            add("pref_key_miuizer_launchericon")
            add("pref_key_miuizer_synced_from_lsposed")
        }

        prefsChanged = SharedPreferences.OnSharedPreferenceChangeListener { sharedPreferences, key ->
            if (AppHelper.remotePrefs == null) return@OnSharedPreferenceChangeListener
            if (key == null) {
                val prefEdit = AppHelper.remotePrefs!!.edit()
                for (remoteKey in AppHelper.remotePrefs!!.all.keys) {
                    prefEdit.remove(remoteKey)
                }
                prefEdit.apply()
                return@OnSharedPreferenceChangeListener
            }
            if (ignoreKeys.contains(key)) return@OnSharedPreferenceChangeListener
            val value = sharedPreferences.all[key] ?: run {
                AppHelper.remotePrefs!!.edit().remove(key).apply()
                return@OnSharedPreferenceChangeListener
            }
            val prefEdit = AppHelper.remotePrefs!!.edit()
            when (value) {
                is Boolean -> prefEdit.putBoolean(key, value)
                is Float -> prefEdit.putFloat(key, value)
                is Int -> prefEdit.putInt(key, value)
                is Long -> prefEdit.putLong(key, value)
                is String -> prefEdit.putString(key, value)
                is Set<*> -> @Suppress("UNCHECKED_CAST") prefEdit.putStringSet(key, value as Set<String>)
            }
            prefEdit.apply()
        }

        prefsChanged?.let { AppHelper.appPrefs!!.registerOnSharedPreferenceChangeListener(it) }
    }

    fun navToSubFragment(
        fragment: Fragment,
        args: Bundle,
        settingsType: AppHelper.SettingsType,
        abType: AppHelper.ActionBarType,
        titleResId: Int,
        contentResId: Int
    ) {
        navToSubFragment(fragment, args, settingsType, abType, resources.getString(titleResId), contentResId)
    }

    fun navToSubFragment(
        fragment: Fragment,
        args: Bundle,
        settingsType: AppHelper.SettingsType,
        abType: AppHelper.ActionBarType,
        title: String,
        contentResId: Int
    ) {
        val currentFragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
        if (currentFragment is PreferenceFragmentBase) {
            currentFragment.openSubFragment(fragment, args, settingsType, abType, title, contentResId)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mainFrag?.let { supportFragmentManager.putFragment(outState, "mainFrag", it) }
        super.onSaveInstanceState(outState)
    }

    @SuppressLint("ApplySharedPref")
    override fun onDestroy() {
        try {
            prefsChanged?.let { AppHelper.appPrefs!!.unregisterOnSharedPreferenceChangeListener(it) }
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        super.onDestroy()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                val fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
                if (fragment == null) {
                    finish()
                } else if (fragment is MainFragment) {
                    finish()
                } else if (fragment is SubFragment) {
                    fragment.finish()
                }
                true
            }
            R.id.resetsettings -> {
                showResetSettingsDialog()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showResetSettingsDialog() {
        AlertDialog.Builder(this)
            .setTitle(R.string.reset_settings)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                AppHelper.appPrefs!!.edit().clear().apply()
                AlertDialog.Builder(this)
                    .setTitle(R.string.reset_settings_done)
                    .setCancelable(true)
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        Process.killProcess(Process.myPid())
                    }
                    .show()
            }
            .setNegativeButton(android.R.string.cancel) { _, _ -> }
            .show()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        return if (keyCode == KeyEvent.KEYCODE_MENU) true else super.onKeyDown(keyCode, event)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        if (grantResults.isEmpty()) {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
            return
        }

        when (requestCode) {
            Helpers.REQUEST_PERMISSIONS_WIFI -> {
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    if (shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_COARSE_LOCATION))
                        Toast.makeText(this, R.string.permission_scan, Toast.LENGTH_LONG).show()
                    else
                        Toast.makeText(this, R.string.permission_permanent, Toast.LENGTH_LONG).show()
                }
            }
            Helpers.REQUEST_PERMISSIONS_BLUETOOTH -> {
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    if (shouldShowRequestPermissionRationale(Manifest.permission.BLUETOOTH_CONNECT))
                        Toast.makeText(this, R.string.permission_scan, Toast.LENGTH_LONG).show()
                    else
                        Toast.makeText(this, R.string.permission_permanent, Toast.LENGTH_LONG).show()
                }
            }
            Helpers.REQUEST_PERMISSIONS_REPORT -> Toast.makeText(this, ":(", Toast.LENGTH_SHORT).show()
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }
}
